pkg/cli/cli.go

```diff
@@ -0,0 +1,273 @@
+limitations under the License.
+*/
+
+package cli
+
+import (
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	internalconfig "sigs.k8s.io/kubebuilder/internal/config"
+	"sigs.k8s.io/kubebuilder/pkg/cli/internal"
+	"sigs.k8s.io/kubebuilder/pkg/plugin"
+)
+
+const (
+	noticeColor = "\033[1;36m%s\033[0m"
+)
+
+// CLI interacts with a command line interface.
+type CLI interface {
+	// Run runs the CLI, usually returning an error if command line configuration
+	// is incorrect.
+	Run() error
+}
+
+// Option is a function that can configure the cli
+type Option func(*cli) error
+
+// cli defines the command line structure and interfaces that are used to
+// scaffold kubebuilder project files.
+type cli struct {
+	commandName    string
+	defProjVersion string
+	cmd            *cobra.Command
+	extraCommands  []*cobra.Command
+	plugins        map[string][]plugin.Base
+}
+
+// New creates a new cli instance.
+func New(opts ...Option) (CLI, error) {
+	c := &cli{
+		commandName:    "kubebuilder",
+		defProjVersion: internalconfig.DefaultVersion,
+		plugins:        map[string][]plugin.Base{},
+	}
+	for _, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := c.initialize(); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Run runs the cli.
+func (c cli) Run() error {
+	return c.cmd.Execute()
+}
+
+// WithCommandName is an Option that sets the cli's root command name.
+func WithCommandName(name string) Option {
+	return func(c *cli) error {
+		c.commandName = name
+		return nil
+	}
+}
+
+// WithDefaultProjectVersion is an Option that sets the cli's default project
+// version. Setting an unknown version will result in an error.
+func WithDefaultProjectVersion(version string) Option {
+	return func(c *cli) error {
+		c.defProjVersion = version
+		return nil
+	}
+}
+
+// WithPlugins is an Option that sets the cli's plugins.
+func WithPlugins(plugins ...plugin.Base) Option {
+	return func(c *cli) error {
+		for _, p := range plugins {
+			ver := p.Version()
+			if _, ok := c.plugins[ver]; !ok {
+				c.plugins[ver] = []plugin.Base{}
+			}
+			c.plugins[ver] = append(c.plugins[ver], p)
+		}
+		return nil
+	}
+}
+
+// WithExtraCommands is an Option that adds extra subcommands to the cli.
+// Adding extra commands that duplicate existing commands results in an error.
+func WithExtraCommands(cmds ...*cobra.Command) Option {
+	return func(c *cli) error {
+		c.extraCommands = append(c.extraCommands, cmds...)
+		return nil
+	}
+}
+
+// initialize initializes the cli.
+func (c *cli) initialize() error {
+	var projectVersion string
+	if internal.Configured() {
+		projectVersion, _ = getProjectVersion()
+	} else {
+		// If the project isn't configured, see if we can find the
+		// project version in a command line flag.
+		projectVersion, _ = c.getBaseFlags()
+	}
+
+	if projectVersion != "" {
+		ps, ok := c.plugins[projectVersion]
```

Comment:
Could we centralize the GetProjectVersion and then, make it global to use in all places?

Reply:
Duplicate of https://github.com/kubernetes-sigs/kubebuilder/pull/1290#discussion_r374980126, lets keep the discussion over there.